[ObjCARC] Move autorelease-to-release conversion to pool pop site instead of converting in place

AZero13 opened the pull request:
Instead of converting objc_autorelease(x) to objc_release(x) in place (which requires checking for uses of x between the autorelease and the pool pop), insert the release just before the matching autoreleasePoolPop. This is safe because objc_autorelease does not change the refcount; it only registers the object for a deferred release at pool drain time. Removing the autorelease and placing an explicit release at the drain point is semantically equivalent.

LumioseSil committed:
[ObjCARC] Move autorelease-to-release conversion to pool pop site instead of converting in place

Instead of converting objc_autorelease(x) to objc_release(x) in place (which requires checking for uses of x between the autorelease and the pool pop), insert the release just before the matching autoreleasePoolPop. This is safe because objc_autorelease does not change the refcount; it only registers the object for a deferred release at pool drain time. Removing the autorelease and placing an explicit release at the drain point is semantically equivalent.

AZero13 commented:
@jroelofs

jroelofs commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -121,6 +121,44 @@ static const Value *FindSingleUseIdentifiedObject(const Value *Arg) {
+      // objects
+      return false;
+
+    default:
```
I recommend fully covering the switch, and getting rid of the `default:` case. Then `-Wswitch` can help alert people to usage sites when they add new entries to the enum.

AZero13 commented:
@rjmccall @jroelofs Ping?

AZero13 commented:
@jroelofs Ping?

LumioseSil committed:
Pre-commit tests (NFC)

SiliconA-Z commented:
@ahatanak Hello?

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -2592,64 +2685,80 @@ void ObjCARCOpt::OptimizeAutoreleasePools(Function &F) {
+          // Mismatched pop.
+          // We can't trust the stack anymore, invalidating optimization for
+          // this block.
+          PoolStack.clear();
```
It seems that the changes made to `ObjCARCOpt::OptimizeAutoreleasePools` aren't needed for the autorelease-to-release conversion. Is that correct? Can you split it out to a separate PR?

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -978,27 +1036,53 @@ void ObjCARCOpt::OptimizeIndividualCallImpl(Function &F, Instruction *Inst,
 
   // objc_autorelease(x) -> objc_release(x) if x is otherwise unused.
-  if (IsAutorelease(Class) && Inst->use_empty()) {
+  if (Class == ARCInstKind::Autorelease && Inst->use_empty()) {
```
Is this change needed for the autorelease-to-release conversion this PR implements? Or is this a separate bug fix?

AZero13 replied:
Restrict this optimization to ordinary objc_autorelease. While converting objc_autoreleaseReturnValue to objc_release is semantically correct, doing so here prevents the existing retainRV/autoreleaseRV optimization from eliminating those pairs. Limiting this optimization to ARCInstKind::Autorelease preserves the more profitable RV optimization.

AZero13 replied:
It was never a bug I meant.

ahatanak replied:
Could you provide a test case demonstrating this? Specifically, one showing that before this change, the code here converted an autoreleaseRV into a release, which prevented the retainRV/autoreleaseRV pairing optimization from firing, and that after this change, the pairing optimization does fire.

AZero13 replied:
Done!

ahatanak replied:
Is `@test_retainRV_autoreleaseRV_pairing_in_pool` the test case that was added to demonstrate this? Reverting `Class == ARCInstKind::Autorelease` back to `IsAutorelease(Class)` doesn't change the IR generated for that test, so it doesn't seem to demonstrate the effect of this change. Could you revert the change if there's no test that demonstrates its effect? Either way, I think this change should be discussed in a separate PR.

AZero13 replied:
> Is `@test_retainRV_autoreleaseRV_pairing_in_pool` the test case that was added to demonstrate this? Reverting `Class == ARCInstKind::Autorelease` back to `IsAutorelease(Class)` doesn't change the IR generated for that test, so it doesn't seem to demonstrate the effect of this change. Could you revert the change if there's no test that demonstrates its effect? Either way, I think this change should be discussed in a separate PR.

I made a mistake and applied it to the wrong block. Fixed!

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -601,6 +606,57 @@ class ObjCARCOpt {
+      // is harmless: the cache is keyed by raw pointer and cleared per
+      // function, and no new Autorelease instructions are created by this pass
+      // that could alias a freed key.
+      ScopeSiblings.push_back(&*I);
```
`ScopeSiblings` only caches autoreleases at `Depth == 0`, so autoreleases inside nested pools are walked but never cached, meaning `FindFollowingAutoreleasePoolPop` re-scans from scratch when one of those is later queried. This reintroduces the quadratic behavior the caching was meant to fix.

Can you cache all autoreleases encountered during the scan?

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -601,6 +606,57 @@ class ObjCARCOpt {
+  for (BasicBlock::iterator I = std::next(AutoreleaseInst->getIterator()),
+                            E = BB->end();
+       I != E; ++I) {
+    ARCInstKind Class = GetBasicARCInstKind(&*I);
```
> A CallOrUser/Call bail-out would make this scan more conservative but wouldn't fix any correctness issue — it would just reduce the number of autoreleases the optimization can handle, for no safety gain.

What if the IR looks like this?
```
  %outer = call ptr @llvm.objc.autoreleasePoolPush()
  call void @llvm.objc.autorelease(ptr %obj)
  %p = call ptr @h()
  call void @llvm.objc.autoreleasePoolPop(ptr %p)
  call void @use(ptr %obj)
  call void @llvm.objc.autoreleasePoolPop(ptr %outer)
```

And `@h` is defined like this:
```
define ptr @h() {
  %p = tail call ptr @lvm.objc.autoreleasePoolPush()
  ret ptr %p
}
```

`%obj` has to be released after the call to `@use`. However, the code in this PR would insert the release before the call to `@use`, right before the first pop.

AZero13 replied:
You are right. So I fixed that!

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -601,6 +606,60 @@ class ObjCARCOpt {
+  Instruction *Result = nullptr;
+
+  SmallVector<SmallVector<Instruction *, 2>, 4> AutoreleasesByDepth(1);
+
```
A few optional simplifications:
 
1. Insert `AutoreleaseInst` into `AutoreleasesByDepth[0]` before the loop starts. It's just another depth-0 sibling like any other autorelease the scan finds, so this lets you remove the special-cased `FollowingPoolPopCache[AutoreleaseInst] = Result;` after the loop.
2. Move the `if (Depth == 0)` check to right after `AutoreleasesByDepth[Depth].clear()`, and return `&*I` immediately from there instead of breaking. This unifies the depth-0 case with the general case, instead of treating it separately.
3. With change 2, once you fall off the end of the loop, every remaining instruction across all of `AutoreleasesByDepth` (including `AutoreleaseInst` itself) is known to have no following pop in this block, so the trailing loop can assign nullptr directly instead of `Result`. You can then drop the `Result` local and return `&*I` / `nullptr` directly instead.
4. In the `AutoreleasepoolPush` branch, assert that `AutoreleasesByDepth[Depth]` is empty when `Depth != (int)AutoreleasesByDepth.size()`. You are reusing a bucket, which has been cleared earlier, rather than growing the vector.

AZero13 replied:
Done!

AZero13 commented:
@ahatanak All is done! I addressed everything else.

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
```diff
@@ -601,6 +606,65 @@ class ObjCARCOpt {
+  auto It = FollowingPoolPopCache.find(AutoreleaseInst);
+  if (It != FollowingPoolPopCache.end()) {
+    // The cached value is a raw pointer to a pool pop. This is safe
+    // because this pass never erases autoreleasePoolPop instructions.
```
This comment isn't accurate. The pass does erase `autoreleasePoolPop` instructions in `OptimizeAutoreleasePools`, but not while `FollowingPoolPopCache` is live and being consulted.

ahatanak commented on llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp:
This changes the release order from LIFO to FIFO, since each converted autorelease is inserted right before the pop in visitation order. I think that's fine as this pass already doesn't preserve pool drain order elsewhere, e.g., the existing `IsAutorelease(Class) && Inst->use_empty()` conversion releases immediately in place.

AZero13 replied:
Yeah, I am keeping it because LIFO requires extra code and extra state. Plus, the few years where function outlining basically broke ARC optimizations everywhere already made the autorelease order unstable.

AZero13 committed:
Pre-commit tests (NFC)

AZero13 committed:
[ObjCARC] Move autorelease-to-release conversion to pool pop site instead of converting in place

Instead of converting objc_autorelease(x) to objc_release(x) in place (which requires checking for uses of x between the autorelease and the pool pop), insert the release just before the matching autoreleasePoolPop. This is safe because objc_autorelease does not change the refcount; it only registers the object for a deferred release at pool drain time. Removing the autorelease and placing an explicit release at the drain point is semantically equivalent.

AZero13 commented:
@ahatanak can you please merge? Thank you
